src/Conversion/ONNXToKrnl/Math/Clip.cpp

```diff
@@ -0,0 +1,104 @@
+//===----------------- Clip.cpp - Lowering Clip Op
+//-------------------------===//
+//
+// Copyright 2019 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file lowers the ONNX Clip Operator to Krnl dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp"
+#include "src/Dialect/ONNX/ONNXShapeHelper.hpp"
+
+using namespace mlir;
+
+//===----------------------------------------------------------------------===//
+// Scalar unary ops for lowering ONNXClipOp
+//===----------------------------------------------------------------------===//
+
+struct ONNXClipOpLowering : public ConversionPattern {
+  ONNXClipOpLowering(MLIRContext *ctx)
+      : ConversionPattern(ONNXClipOp::getOperationName(), 1, ctx) {}
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    Location loc = op->getLoc();
+    Value input = operands[0];
+    Value min = operands[1];
+    Value max = operands[2];
+
+    // Insert an allocation and deallocation for the result of this operation.
+    auto memRefType = convertToMemRefType(*op->result_type_begin());
+
+    Value alloc;
+    bool insertDealloc = checkInsertDealloc(op);
+
+    if (hasAllConstantDimensions(memRefType))
+      alloc = insertAllocAndDealloc(memRefType, loc, rewriter, insertDealloc);
+    else
+      alloc = insertAllocAndDealloc(
+          memRefType, loc, rewriter, insertDealloc, input);
+
+    SmallVector<Value, 4> loopIVs;
+    // Only create krnl.iterate if one of the operands is not scalar tensor.
+    if (!hasAllScalarValues(operands)) {
+      // Create iterateOp & get block within iterate op.
+      BuildKrnlLoop loops(rewriter, loc, memRefType.getRank());
+      loops.createDefineAndIterateOp(input);
+      Block *iterationBlock = loops.getIterateBlock();
+
+      // Insert instructions inside the KernelIterateOp body.
+      rewriter.setInsertionPointToStart(iterationBlock);
```

Comment:
I think you might need to use an insertion point guard so that the insertion point is restored to the previous one when this block ends.

Reply:
I copied the code from Elementwise.cpp; I really don't know much how the insertion point is set before entering such functions, I think it never need to be reset for coherent use after the function is over.

Here, the use is fine: if there is a loop, emit the loop and set insertion point inside; if there is no loop, insert where we would have put the loops.

Reply:
As long as it's correct I'm ok with it :) I though it would be worth another glance in case there is something wrong with it. Thanks for double checking.